test/framework/image_preloading.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/cluster-api/test/framework"
+	. "sigs.k8s.io/cluster-api/test/framework/ginkgoextensions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func LoadImagesFunc(ctx context.Context) func(clusterProxy framework.ClusterProxy) {
+	sourceFile := os.Getenv("DOCKER_IMAGE_TAR")
+	Expect(sourceFile).ToNot(BeEmpty(), "DOCKER_IMAGE_TAR must be set")
+
+	loader := imagePreloader{
+		sourceFile: sourceFile,
+	}
+
+	return func(clusterProxy framework.ClusterProxy) {
+		loader.ImagesToCluster(ctx, clusterProxy)
+	}
+}
+
+type imagePreloader struct {
+	sourceFile string
+}
+
+// ImagesToCluster deploys a privileged daemonset and uses it to stream-load container images.
+func (loader *imagePreloader) ImagesToCluster(ctx context.Context, clusterProxy framework.ClusterProxy) {
```

Comment:
```suggestion
func (loader *imagePreloader) loadImagesToCluster(ctx context.Context, clusterProxy framework.ClusterProxy) {
```
I don't know loader.ImagesToCluster reads strange. I would have probably also dropped the struct. Looks like we don't really need it